Make it possible to get the collection for an object

tmadlener opened the pull request:
BEGINRELEASENOTES
- Add `Frame::get` overloads that take a podio generated object handle as argument and return the collection this object belongs to *if the collection is known to the Frame*.

ENDRELEASENOTES

See the discussion in https://github.com/key4hep/k4FWCore/issues/311 for a rationale of this functionality

The main question at this point is whether the existing `getName` functionality that takes a `collectionID` should be deprecated and removed eventually.

tmadlener committed:
Make it possible to get the collection for an object

m-fila commented on include/podio/Frame.h:
```diff
@@ -199,6 +199,29 @@ class Frame {
+  ///
+  /// @returns       A const reference to the collection if it is available or to
+  ///                an empty (static) collection
+  template <CollectionType CollT>
+  const CollT& get(const CollT::value_type& object) const;
```
```suggestion
  template <CollectionType CollT>
  const CollT& get(const typename CollT::value_type& object) const;
```
Adding a typename here (and in definition below) shouldn't hurt (although we already `value_type` to be type in the concept)

tmadlener replied:
Interesting that this seems to work without `typename` for gcc11, but not for gcc13. Fixed.

m-fila commented on include/podio/Frame.h:
```diff
@@ -199,6 +199,29 @@ class Frame {
+  ///
+  /// @returns A const pointer to a collection if it is available or a nullptr
+  ///          if it is not
+  template <ObjectType O>
+  inline const podio::CollectionBase* get(const O& object) const;
```
Do we want to have this for compatibility with other `get`s in frame? Otherwise, I think the objects already have typedef to their collection so in principle we could use only the overload returning concrete collection

tmadlener committed:
Add tests for ObjectType concept and fix links

Link defines the value_type type member which should be object_type for
conformancy with other generated handles.

tmadlener committed:
Add typename to fix compilation error

tmadlener committed:
Return a typed pointer if possible

tmadlener committed:
Add unittests for python bindings
